WordPress/gutenberg
# [WIP] Style engine: add elements and states to the backend

## What?

Updates the Style Engine to add support for:
- `elements` styles - limited to `link` and `button`.
- individual element "states" - e.g., `a:hover`, `focus` and so on (limited to `link` only).

Initially only `link` elements are supported with the `:hover` pseudo selector. However the system is open to extension so in the future it will be easy to add support for more selectors.

Background context: https://github.com/WordPress/gutenberg/pull/41383#issuecomment-1149931908

Related to:
- https://github.com/WordPress/gutenberg/pull/40987


```php

wp_style_engine_generate(
    array(
        'elements' => array(
            'link' => array(
                'color'  => array(
                    'text'       => '#fff',
                    'background' => '#000',
                ),				
                ':hover' => array(
                    'color' => array(
                        'text'       => '#000',
                        'background' => '#fff',
                    ),
                ),				
            ),
        ),
    )
, array( 'selector' => '.la-sinistra' ) );


/*

Returns: 

array(
    'css' => '.la-sinistra a { color: #fff; background-color: #000; } .la-sinistra a:hover { color: #000; background-color: #fff; }',
)

*/
```


## Why?

Theme developers and user are requesting the ability to control interactivity states. By adding support to the Style Engine we prepare the ground for the code that handles both editor and front end styles to output the correct CSS based on the `theme.json`.

In the future this will also open the door to a UI for configuring interactivity states. 

## How?

The style engine will now iterate over `elements` key of the block styles and generate the appropriate CSS based on the rules found within those definitions.

In addition, any pseudo selectors found within `elements` will be generated for the given element, having first been checked against an "allowed list" (to avoid things like `h1:hover` which is not a best practice and should be avoided).

## Testing Instructions

Follow the test instructions over at https://github.com/WordPress/gutenberg/pull/40987. 

There should be no change in the way we output link styles in `elements.php`

In relation to the new "states", e.g., `:hover`, currently this is only backend support and adding rules to `theme.json` will have no effect. This PR updates the style engine so check the unit tests make sense and cover the key scenarios.

```bash
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/packages/style-engine/phpunit/class-wp-style-engine-test.php
```
